### ISSUE TITLE
Handling large data- Table Paginating

### ISSUE THREAD

**Issue by A7medMaher** [bosskmk/pluto_grid#105]
Hello Again.

Any plans to add paginating when dealing with large data ?  do you suggest an alternative method for large data handling?

**Comment by bosskmk** [comment]
Pagination is a must-have feature. There are plans to add. 
I think I will do it when the filtering I'm currently doing is over.

**Comment by jsanchez-98** [comment]
Any news on that or should I just give up on this plugin.

**Comment by fehernyul** [comment]
Is there any progression on this feature?

**Comment by Minsamin** [comment]
Hello, are you working on the pagination part? Any ETA ? I want to use this library in my project.

**Comment by bosskmk** [comment]
I'm thinking about how to do it, but I can't because I don't have much time.

**Comment by amasahara** [comment]
Hope that you can do it soon because it very very necessary for big data with render. long time for waiting

**Comment by bosskmk** [comment]
I am working on pagination.
If the existing createFooter callback returns a Pagination widget, it causes pagination on the client side.
One Pagination widget is provided by default and will allow users to customize it.
It is expected that it will be completed and distributed by the end of next week at the latest tomorrow at the latest.

**Comment by amasahara** [comment]
thanks for great package, hope it will be soon

**Comment by bosskmk** [comment]
Version 2.4.0 has been released.

example demo
https://weblaze.dev/pluto_grid/build/web/#feature/row-pagination

**Comment by bosskmk** [comment]
I found a bug where if I filter on a paged list and then release it again, the list doesn't come back.

There may be bugs due to insufficient testing. (to be fixed in next release)

If you find any other bugs, please report them to me.

**Comment by Minsamin** [comment]
Can we use this with lazy data loading?
Like, get the particular page data from api, then put data in dataTable.
For every page, it will first fetch then show in dataTable...

**Comment by bosskmk** [comment]
@Minsamin 
That feature should be added to the server-side rendering feature.
Or you have to implement it yourself.
Most APIs can be called from StateManager that is passed as an argument to the onLoaded callback function when creating a grid.
If you use this, you will be able to implement it yourself.

**Comment by Minsamin** [comment]
I think, if PlutoPagination has buttonPress callback(first, last, next, Prev) for user, and any method to show loading while fetching data from server, that will be helpful.

**Comment by bosskmk** [comment]
@Minsamin 
First of all, thanks for your comments.

The features you mentioned are currently available.
```dart
stateManager.setPage()
stateManager.setPageSize()
stateManager.page
stateManager.pageSize
stateManager.totalPage
stateManager.setShowLoading()
```

However, additional callback events are required for filtering and sorting.

I'll try to add server-side pagination to the next release.

**Comment by amasahara** [comment]
stateManager.totalPage is getter only which way i can use it. Could you make an example for it

**Comment by bosskmk** [comment]
@amasahara yes it is. `stateManager.totalPage is getter.

**Comment by amasahara** [comment]
So if i want so data like advance datatable i will be waiting ultil next update or another way to do it

**Comment by jamolina2021** [comment]
Hi @bosskmk,
Any progress on server-side pagination? 
Thanks for such a wonderful datagrid. 
Regards,
Jose

**Comment by tuananh432** [comment]
@bosskmk how can i set the page change event?

**Comment by reke592** [comment]
hi, is there a way to manually declare the paging process? I'm trying to figure out how to use the paginator, below is my sample backend response.
```js
{
  data: [...],
  first: "/api/v1/models?page=1&limit=50",
  last: "/api/v1/models?page=261&limit=50",
  next: "/api/v1/models?page=4&limit=50",
  previous: "/api/v1/models?page=2&limit=50",
  page_no: 3,
  page_size: 50,
  total_page: 261,
  total_rows: 13029
}
```

what we are trying to achieve is something like this in UI
![image](https://user-images.githubusercontent.com/17094894/155841800-2f952a9d-3f2f-4c9b-bce0-529303f93d85.png)

**Comment by a7mdragab** [comment]
> I think, if PlutoPagination has buttonPress callback(first, last, next, Prev) for user, and any method to show loading while fetching data from server, that will be helpful.
> 
> reference https://dev-owl.github.io/advanced_datatable/#/

Any updates about this?

**Comment by acamenhas** [comment]
Hello,

Unfortunately from what I see the async pagination is not ready yet.

It's something that is not very complex:

1- page change event (know what page we are)
2- possibility to show a spinner
3- api call
4- refresh data
5- hide spinner

**Comment by bosskmk** [comment]
Yes, I am planning to provide a widget for asynchronous pagination.
I just haven't started doing this yet.
I'll start this up and let you know when I'm done.
thank you

**Comment by bosskmk** [comment]
I am not yet developing the asynchronous pagination feature, but users can implement server side pagination as shown below.
(I have plans to add features in the future)

Sorting and filtering require minor modifications to the current event part.

Currently, when a user's sorting or filtering event occurs, it is processed as local data. In server-side pagination, only the user's event is generated and the data must be imported as the event information.

```dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:pluto_grid/pluto_grid.dart';

import '../dummy_data/development.dart';

class EmptyScreen extends StatefulWidget {
  static const routeName = 'empty';

  const EmptyScreen({Key? key}) : super(key: key);

  @override
  _EmptyScreenState createState() => _EmptyScreenState();
}

class _EmptyScreenState extends State<EmptyScreen> {
  final List<PlutoColumn> columns = DummyData.textColumns(5);

  final List<PlutoRow> rows = [];

  final List<PlutoRow> fakeFetchedRows = [];

  late PlutoGridStateManager stateManager;

  final int pageSize = 30;

  @override
  void initState() {
    super.initState();

    /// 100 data generation for fake server data.
    fakeFetchedRows.addAll(
      DummyData.rowsByColumns(length: 100, columns: columns),
    );
  }

  Future<_PaginationData> fetch(
    _PaginationParams params,
  ) async {
    /// Get data from server.
    final start = (params.page - 1) * pageSize;
    final end = min(start + pageSize, fakeFetchedRows.length);
    final fetchRows = fakeFetchedRows.getRange(start, end);
    await Future.delayed(const Duration(seconds: 1));

    /// Change server data to Map<String, dynamic> type.
    return Future.value(
      _PaginationData(
        totalPage: (fakeFetchedRows.length / pageSize).ceil(),
        rows: fetchRows
            .map(
              (e) => e.cells.map<String, dynamic>(
                (key, value) => MapEntry(key, value.value),
              ),
            )
            .toList(),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: Container(
          padding: const EdgeInsets.all(15),
          child: PlutoGrid(
            columns: columns,
            rows: rows,
            onChanged: (PlutoGridOnChangedEvent event) {
              print(event);
            },
            onLoaded: (PlutoGridOnLoadedEvent event) {
              stateManager = event.stateManager;
            },
            configuration: const PlutoGridConfiguration(),
            createFooter: (s) => PlutoPaginationServerSide(s, fetch: fetch),
          ),
        ),
      ),
    );
  }
}

class PlutoPaginationServerSide extends StatefulWidget {
  const PlutoPaginationServerSide(
    this.stateManager, {
    required this.fetch,
    super.key,
  });

  final PlutoGridStateManager stateManager;

  final Future<_PaginationData> Function(
    _PaginationParams params,
  ) fetch;

  @override
  State<PlutoPaginationServerSide> createState() =>
      _PlutoPaginationServerSideState();
}

class _PlutoPaginationServerSideState extends State<PlutoPaginationServerSide> {
  PlutoGridStateManager get stateManager => widget.stateManager;

  final int initialPage = 1;

  int totalPage = 0;

  @override
  void initState() {
    super.initState();

    WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
      setPage(initialPage);
    });
  }

  Future<void> setPage(int page) async {
    stateManager.setShowLoading(true, level: PlutoGridLoadingLevel.rows);

    final data = await widget.fetch(
      _PaginationParams(page: page),
    );

    if (totalPage != data.totalPage) setState(() => totalPage = data.totalPage);

    stateManager.refRows.clearFromOriginal();
    stateManager.insertRows(
      0,
      data.rows
          .map(
            (e) => PlutoRow(
              cells: {
                for (var v in e.entries) v.key: PlutoCell(value: v.value)
              },
            ),
          )
          .toList(),
    );

    stateManager.setShowLoading(false);
  }

  @override
  Widget build(BuildContext context) {
    return _PaginationWidget(
      iconColor: stateManager.style.iconColor,
      disabledIconColor: stateManager.style.disabledIconColor,
      activatedColor: stateManager.style.activatedBorderColor,
      iconSize: stateManager.style.iconSize,
      height: stateManager.footerHeight,
      initialPage: initialPage,
      totalPage: totalPage,
      setPage: setPage,
    );
  }
}

class _PaginationData {
  _PaginationData({
    required this.totalPage,
    required this.rows,
  });

  final int totalPage;

  final List<Map<String, dynamic>> rows;
}

class _PaginationParams {
  _PaginationParams({
    required this.page,
  });

  final int page;
}

class _PaginationWidget extends StatefulWidget {
  const _PaginationWidget({
    required this.iconColor,
    required this.disabledIconColor,
    required this.activatedColor,
    required this.iconSize,
    required this.height,
    this.initialPage = 1,
    required this.totalPage,
    required this.setPage,
  });

  final Color iconColor;

  final Color disabledIconColor;

  final Color activatedColor;

  final double iconSize;

  final double height;

  final int initialPage;

  final int totalPage;

  final void Function(int page) setPage;

  @override
  State<_PaginationWidget> createState() => _PaginationWidgetState();
}

class _PaginationWidgetState extends State<_PaginationWidget> {
  @override
  void initState() {
    super.initState();

    _page = widget.initialPage;
  }

  int get totalPage => widget.totalPage;

  int _page = 1;

  late double _maxWidth;

  final _iconSplashRadius = PlutoGridSettings.rowHeight / 2;

  bool get _isFirstPage => _page < 2;

  bool get _isLastPage => _page > totalPage - 1;

  /// maxWidth < 450 : 1
  /// maxWidth >= 450 : 3
  /// maxWidth >= 550 : 5
  /// maxWidth >= 650 : 7
  int get _itemSize {
    final countItemSize = ((_maxWidth - 350) / 100).floor();

    return countItemSize < 0 ? 0 : min(countItemSize, 3);
  }

  int get _startPage {
    final itemSizeGap = _itemSize + 1;

    var start = _page - itemSizeGap;

    if (_page + _itemSize > totalPage) {
      start -= _itemSize + _page - totalPage;
    }

    return start < 0 ? 0 : start;
  }

  int get _endPage {
    final itemSizeGap = _itemSize + 1;

    var end = _page + _itemSize;

    if (_page - itemSizeGap < 0) {
      end += itemSizeGap - _page;
    }

    return end > totalPage ? totalPage : end;
  }

  List<int> get _pageNumbers {
    return List.generate(
      _endPage - _startPage,
      (index) => _startPage + index,
    );
  }

  void _firstPage() {
    _movePage(1);
  }

  void _beforePage() {
    int beforePage = _page - 1 + (_itemSize * 2);

    if (beforePage < 1) {
      beforePage = 1;
    }

    _movePage(beforePage);
  }

  void _nextPage() {
    int nextPage = _page + 1 + (_itemSize * 2);

    if (nextPage > totalPage) {
      nextPage = totalPage;
    }

    _movePage(nextPage);
  }

  void _lastPage() {
    _movePage(totalPage);
  }

  void _movePage(int page) {
    setState(() {
      _page = page;
      widget.setPage(page);
    });
  }

  ButtonStyle _getNumberButtonStyle(bool isCurrentIndex) {
    return TextButton.styleFrom(
      disabledForegroundColor: Colors.transparent,
      shadowColor: Colors.transparent,
      padding: const EdgeInsets.fromLTRB(5, 0, 0, 10),
      backgroundColor: Colors.transparent,
    );
  }

  TextStyle _getNumberTextStyle(bool isCurrentIndex) {
    return TextStyle(
      fontSize: isCurrentIndex ? widget.iconSize : null,
      color: isCurrentIndex ? widget.activatedColor : widget.iconColor,
    );
  }

  Widget _makeNumberButton(int index) {
    var pageFromIndex = index + 1;

    var isCurrentIndex = _page == pageFromIndex;

    return TextButton(
      onPressed: () {
        _movePage(pageFromIndex);
      },
      style: _getNumberButtonStyle(isCurrentIndex),
      child: Text(
        pageFromIndex.toString(),
        style: _getNumberTextStyle(isCurrentIndex),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (_, size) {
        _maxWidth = size.maxWidth;

        final Color iconColor = widget.iconColor;

        final Color disabledIconColor = widget.disabledIconColor;

        return Container(
          width: _maxWidth,
          height: widget.height,
          alignment: Alignment.center,
          child: SingleChildScrollView(
            scrollDirection: Axis.horizontal,
            child: Row(
              children: [
                IconButton(
                  onPressed: _isFirstPage ? null : _firstPage,
                  icon: const Icon(Icons.first_page),
                  color: iconColor,
                  disabledColor: disabledIconColor,
                  splashRadius: _iconSplashRadius,
                  mouseCursor: _isFirstPage
                      ? SystemMouseCursors.basic
                      : SystemMouseCursors.click,
                ),
                IconButton(
                  onPressed: _isFirstPage ? null : _beforePage,
                  icon: const Icon(Icons.navigate_before),
                  color: iconColor,
                  disabledColor: disabledIconColor,
                  splashRadius: _iconSplashRadius,
                  mouseCursor: _isFirstPage
                      ? SystemMouseCursors.basic
                      : SystemMouseCursors.click,
                ),
                ..._pageNumbers.map(_makeNumberButton).toList(),
                IconButton(
                  onPressed: _isLastPage ? null : _nextPage,
                  icon: const Icon(Icons.navigate_next),
                  color: iconColor,
                  disabledColor: disabledIconColor,
                  splashRadius: _iconSplashRadius,
                  mouseCursor: _isLastPage
                      ? SystemMouseCursors.basic
                      : SystemMouseCursors.click,
                ),
                IconButton(
                  onPressed: _isLastPage ? null : _lastPage,
                  icon: const Icon(Icons.last_page),
                  color: iconColor,
                  disabledColor: disabledIconColor,
                  splashRadius: _iconSplashRadius,
                  mouseCursor: _isLastPage
                      ? SystemMouseCursors.basic
                      : SystemMouseCursors.click,
                ),
              ],
            ),
          ),
        );
      },
    );
  }
}
```

**Comment by bosskmk** [comment]
Released PlutoGrid 5.3.0

Refer to below link.
https://github.com/bosskmk/pluto_grid/issues/601
